numpy/core/src/multiarray/mapping.c

```diff
@@ -235,29 +206,76 @@ prepare_index(PyArrayObject *self, PyObject *index,
 
         if (make_tuple) {
             /* We want to interpret it as a tuple, so make it one */
-            index = PySequence_Tuple(index);
-            if (index == NULL) {
-                return -1;
-            }
+            index_as_tuple = PySequence_Tuple(index);
         }
     }
 
-    /* If the index is not a tuple, handle it the same as (index,) */
-    if (!PyTuple_CheckExact(index)) {
-        obj = index;
-        index_ndim = 1;
+    /* If the index is not a tuple, conver it into (index,) */
+    if (!make_tuple && !PyTuple_CheckExact(index)) {
+        index_as_tuple = Py_BuildValue("(O)", index);
```

Comment:
This is the more inefficient line. If this is unacceptable, then some extra arguments could enable/disable this behaviour. I would like this to stay though, so that it can be exposed python-side